dev/migrate.ps1

```diff
@@ -25,36 +27,62 @@ if ($all -or $postgres -or $mysql -or $sqlite) {
   }
 }
 
+function Get-UserSecrets {
+  # The dotnet cli command sometimes adds //BEGIN and //END comments to the output, Where-Object removes comments
+  # to ensure a valid json
+  return dotnet user-secrets list --json --project "$currentDir/../src/Api" | Where-Object { $_ -notmatch "^//" } | ConvertFrom-Json
+}
+
 if ($all -or $mssql) {
-  function Get-UserSecrets {
-    # The dotnet cli command sometimes adds //BEGIN and //END comments to the output, Where-Object removes comments
-    # to ensure a valid json
-    return dotnet user-secrets list --json --project ../src/Api | Where-Object { $_ -notmatch "^//" } | ConvertFrom-Json
-  }
+  if ($all -or !$test) {
+    if ($selfhost) {
+      $msSqlConnectionString = $(Get-UserSecrets).'dev:selfHostOverride:globalSettings:sqlServer:connectionString'
+      $envName = "self-host"
+    } else {
+      $msSqlConnectionString = $(Get-UserSecrets).'globalSettings:sqlServer:connectionString'
+      $envName = "cloud"
+    }
 
-  if ($selfhost) {
-    $msSqlConnectionString = $(Get-UserSecrets).'dev:selfHostOverride:globalSettings:sqlServer:connectionString'
-    $envName = "self-host"
-  } else {
-    $msSqlConnectionString = $(Get-UserSecrets).'globalSettings:sqlServer:connectionString'
-    $envName = "cloud"
+    Write-Host "Starting Microsoft SQL Server Migrations for $envName"
+    dotnet run --project ../util/MsSqlMigratorUtility/ "$msSqlConnectionString"
   }
 
-  Write-Host "Starting Microsoft SQL Server Migrations for $envName"
-
-  dotnet run --project ../util/MsSqlMigratorUtility/ "$msSqlConnectionString"
+  if ($all -or $test) {
+    $testMsSqlConnectionString = $(Get-UserSecrets).'databases:3:connectionString'
+    if ($testMsSqlConnectionString) {
+      $testEnvName = "test databases"
+      Write-Host "Starting Microsoft SQL Server Migrations for $testEnvName"
+      dotnet run --project ../util/MsSqlMigratorUtility/ "$testMsSqlConnectionString"
+    } else {
+      Write-Host "Connection string for a test MSSQL database not found in secrets.json!"
+    }
+  }
 }
 
-$currentDir = Get-Location
-
-Foreach ($item in @(@($mysql, "MySQL", "MySqlMigrations"), @($postgres, "PostgreSQL", "PostgresMigrations"), @($sqlite, "SQLite", "SqliteMigrations"))) {
+Foreach ($item in @(
+    @($mysql, "MySQL", "MySqlMigrations", "mySql", 2),
+    @($postgres, "PostgreSQL", "PostgresMigrations", "postgreSql", 0),
+    @($sqlite, "SQLite", "SqliteMigrations", "sqlite", 1)
```

Comment:
The test attributes (`DatabaseData`, `DatabaseTheory`) allow these providers to be defined in any order, but this script assumes that postgres is 0, sqlite is 1, mySql is 2 and mssql is 3. That won't be the case for everyone, depending on how they've set up their user secrets.

You could add logic here iterating over the user secrets to find the index for each database type, but it might make more sense to update the structure of the user secrets themselves to adopt a more conventional key-value format, like `test:database:mysql:connectionString`. That would also require updating the test attributes, so it's a little in the weeds, but it would make sure this Just Works for everyone.

Reply:
We can't change the format to be `test:database:mysql:connectionString` where `mysql` is treated as the `SupportedDatabaseProviders` enum. This would limit you from being able to setup the same provider multiple times which these tests were explicitly made to support. See mariadb being [added to the pipeline here](https://github.com/bitwarden/server/pull/4989). I would also want it to be able to run against different versions of the same database. If we want to change it to where it's `databases:[pretty name]:connectionString` but then there is still `databases:[pretty name]:type` then I'd be fine with that but that doesn't really solve making this easier. 

Also, for what it's worth I am planning a pretty big overhaul to these tests when xUnit v3 releases. The main thing being it's added a `AssemblyFixtureAttribute` which:
> can be applied at the assembly-level to add an assembly-wide fixture. Fixtures at this level are created before any test is run, and cleaned up after all tests have finished running.
This is what I have been waiting for to make auto instrumentation possible, so the hope is that we will automatically create a database and run migrations for all possible database types. And when it's not able to run on certain types it will show those types as skipped so that it's a bit more obvious that you are missing some configuration.

Reply:
That makes sense! I hadn't considered multiple providers of the same type. In the meantime, maybe we just keep the fixed array index numbers in the script, or add some logic to identify which type each array entry is?